File: bindings/java/src/main/java/org/indunet/robot/bus/TypedServiceClient.java
package org.indunet.robot.bus;

import com.google.protobuf.MessageLite;

/** Service client that encodes requests and decodes responses. */
public final class TypedServiceClient<Req extends MessageLite, Resp extends MessageLite>
        implements AutoCloseable {
    private final ServiceClient inner;
    private final Class<Req> requestType;
    private final Class<Resp> responseType;

    TypedServiceClient(ServiceClient inner, Class<Req> requestType, Class<Resp> responseType) {
        this.inner = inner;
        this.requestType = requestType;
        this.responseType = responseType;
    }

    public String serviceName() {
        return inner.serviceName();
    }

    public Class<Req> requestType() {
        return requestType;
    }

    public Class<Resp> responseType() {
        return responseType;
    }

    public Resp call(Req request) {
        return call(request, -1.0);
    }

    public Resp call(Req request, double timeoutSecs) {
        if (request == null) {
            throw new NullPointerException("request");
        }
        if (!requestType.isInstance(request)) {
            throw new IllegalArgumentException(
                    "client for "
                            + requestType.getSimpleName()
                            + " got "
                            + request.getClass().getSimpleName());
        }
        byte[] raw = inner.call(ProtoCodec.encode(request), timeoutSecs);
        Resp reply = ProtoCodec.tryParse(responseType, raw);
        if (reply == null) {
            throw new IllegalStateException(
                    "service " + serviceName() + " response decode failed for " + responseType.getSimpleName());
        }
        return reply;
    }

    @Override
    public void close() {
        inner.close();
    }

    @Override
    public String toString() {
        return "TypedServiceClient{service="
                + serviceName()
                + ", request="
                + requestType.getSimpleName()
                + ", response="
                + responseType.getSimpleName()
                + '}';
    }
}

File: bindings/android/src/main/kotlin/org/indunet/robot/bus/TypedServiceClient.kt
package org.indunet.robot.bus

import com.google.protobuf.MessageLite

/** Service client that encodes requests and decodes responses. */
class TypedServiceClient<Req : MessageLite, Resp : MessageLite> internal constructor(
    private val inner: ServiceClient,
    private val requestType: Class<Req>,
    private val responseType: Class<Resp>,
) : AutoCloseable {
    fun serviceName(): String = inner.serviceName()

    fun requestType(): Class<Req> = requestType

    fun responseType(): Class<Resp> = responseType

    @JvmOverloads
    fun call(request: Req, timeoutSecs: Double = -1.0): Resp {
        requireNotNull(request) { "request" }
        if (!requestType.isInstance(request)) {
            throw IllegalArgumentException(
                "client for ${requestType.simpleName} got ${request.javaClass.simpleName}",
            )
        }
        val raw = inner.call(ProtoCodec.encode(request), timeoutSecs)
        return ProtoCodec.tryParse(responseType, raw)
            ?: throw IllegalStateException(
                "service ${serviceName()} response decode failed for ${responseType.simpleName}",
            )
    }

    override fun close() {
        inner.close()
    }

    override fun toString(): String =
        "TypedServiceClient{service=${serviceName()}, request=${requestType.simpleName}, response=${responseType.simpleName}}"
}
